sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecisionSuite.scala

```diff
@@ -276,9 +276,9 @@ class DecimalPrecisionSuite extends AnalysisTest with BeforeAndAfter {
       val a = AttributeReference("a", DecimalType(3, -10))()
       val b = AttributeReference("b", DecimalType(1, -1))()
       val c = AttributeReference("c", DecimalType(35, 1))()
-      checkType(Multiply(a, b), DecimalType(5, -11))
-      checkType(Multiply(a, c), DecimalType(38, -9))
-      checkType(Multiply(b, c), DecimalType(37, 0))
+      checkType(Multiply(a, b), DecimalType(16, 0))
```

Comment:
In this pr, I changed all BinaryArithmetic with negative scale. These test changes you see are actually the legacy issues since we support negative scale, not related https://github.com/apache/spark/pull/36698. I really can not understand why `decimal(38, -9)` can appear in Spark, it means overflow.
The regression of  `IntegralDivide` is caused by https://github.com/apache/spark/pull/36698. We can move code to `IntegralDivide` if we want to narrow the change.

Reply:
Sorry about the long explanation. I am not sure how to make it any shorter. 

By definition a decimal multiply will add the scales. https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html there is no special case for negative scale vs positive scale. It adds them.

```
scala> val lhs = BigDecimal("123E10")
lhs: scala.math.BigDecimal = 1.23E+12

scala> lhs.scale
res0: Int = -10

scala> val rhs = BigDecimal("9E1")
rhs: scala.math.BigDecimal = 9E+1

scala> rhs.scale
res1: Int = -1

scala> val ret = rhs * lhs
ret: scala.math.BigDecimal = 1.107E+14

scala> ret.scale
res2: Int = -11
```

By definition the resulting precision of a multiply can at most be `lhs.precision + rhs.precision + 1`. Again there is no call out for negative scale or positive scale. This is in the SQL standard and also how decimal math works. The problem is that you are resetting the values to have a scale of 0, and effectively changing the precision of the result. This does not follow the standard rules for decimal operations. You are effectively increasing the precision of the value. You get an equivalent answer, but the cost of storing the result is now much greater. So much so that many results will not fit and result in an overflow when they could have fit if the result had a negative scale.

I am -1 on this change as is.  It is wrong to modify multiply in this way.  

Divide is technically wrong in a number of ways with negative scale decimal values even before 3.4.0. The scale of a divide is LHS.scale - RHS.scale. This includes negative scale values. There is technically a great deal of scale loss when doing a decimal divide. Spark follows Hive and most other SQL implementations by having at least a scale of 6 in the output of the divide, and some complicated math to calculate the output precision that goes with it. But that does not deal with negative values cleanly. It would be really nice to know what Hive does in these cases, or what MsSQL does, or really anything that supports negative scale. Do they all return errors? What type do they return if it is not an error? It appears that the SQL spec itself has the bug in it, and I am not inclined to "fix" the spec without some understanding of what others are doing too.

The reason `IntegralDivide` is failing now, where it didn't before, is because the output result was never checked for overflow before being converted into a Long.  `IntegralDivide` returns a Long. The overflow check in 3.4.0 that happens after the divide is totally skipped in previous versions because the output is a long.  Look at the rule in DecimalPrecision

https://github.com/apache/spark/blob/fbbcf9434ac070dd4ced4fb9efe32899c6db12a9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala#L192-L202

If the output type is not a decimal do nothing.  If you want to have `IntegralDivide` act the same as before you need to just remove the overflow check in it. I am not sure if that is a good thing or not.

Reply:
Not sure what you found. I think the reason why Divide and IntegralDivide fail is simple. SQL strandard does not allow negative scale, but we use its definition formula to calculate the result precision and scale. Then the result precision can be nagetive which is unexpected. So I think other binary arithmetic also should not follow if scale is negative.

Reply:
What you are saying is that you want to remove negative scale decimal values by turning them into 0 scale decimal values because the SQL standard does not allow for negative scale decimal, but only for binary math expressions.  Why not everywhere? Why should abs still return a negative scale decimal? Why should cast allow us to cast to a negative scale decimal? How is that better than removing support for negative scale decimal?

Reply:
I think this behavior follows what `spark.sql.legacy.allowNegativeScaleOfDecimal.enabled` said:
https://github.com/apache/spark/blob/a205e97ad9ae7894b2ec27e5253da21c4500fc8c/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L3308-L3313

As I mentioned, I'm ok to only fix the regression with `IntegralDivide` by changing negative scale to 0, and leave others. The cost is huge to change the behavior of all expression which support decimal type.

Reply:
I am fine with this going in just for IntegralDivide, and assuming that others are okay with it for regular Divide too. I think the others should be left as is, at least without more discussion from others.

Reply:
@cloud-fan what do you think ? only fix the regression and leave others as is.